### ISSUE TITLE
MACD Result inaccurate?

### ISSUE THREAD

**Issue by Gonoshift** [bennycode/trading-signals#214]
I saw in issue #165  that @bennycode was referring to [this page](https://tulipindicators.org/macd) to check MACD calculations against.
Problem is that if I use the same setup, the output is not even close to the one shown in the aforementioned page table. 

Am I missing something here?

My test code:

```
import {EMA, MACD} from 'trading-signals';

const macd = new MACD({longInterval: 5, shortInterval: 2, signalInterval: 9, indicator: EMA});
macd.update(81.59);
macd.update(81.06);
macd.update(82.87);
macd.update(83);
macd.update(83.61);

console.log("Signal:" +parseFloat(result.signal.valueOf()).toFixed(4));
console.log("Macd:" +parseFloat(result.macd.valueOf()).toFixed(4));
console.log("Histogram:" +parseFloat(result.histogram.valueOf()).toFixed(4));
```

Result:
```
Signal: 0.2416
Macd: 0.6178
Histogram: 0.3762
```

Expected Result:
```
Signal: 0.62
Macd: 0.62
Histogram: 0.0
```

**Comment by bennycode** [comment]
Hi @Gonoshift, to verify the compatibility with "Tulip Indicators", I have used the same test input as listed in their [Example Calculation](https://tulipindicators.org/macd):

https://github.com/bennycode/trading-signals/blob/22328ea32ceb97f07e5f951534e66daf392cdb09/src/MACD/MACD.test.ts#L9-L25

You should avoid using `parseFloat` because I am using [big.js](https://github.com/MikeMcl/big.js/), which has a higher accuracy. You can directly call `toFixed` on `signal`, `macd` and `histogram`. If you use `result.macd.toFixed(2)`, then the "Macd" result should also be 0.62 on your end. Can you check that? 

My tests don't cover "Signal" and "Histogram", so there may be a difference. That needs to be verified using the Tulip implementation: https://github.com/TulipCharts/tulipindicators/blob/cffa15e389b7a0c472588f22fa326a78bd734391/indicators/macd.c#L68-L83

**Comment by Gonoshift** [comment]
Heya @bennycode , thank you for the reply, I've tried both approaches to check approx errors:

```
console.log("Signal:" +result.signal.toFixed(4));
console.log("Macd:" +result.macd.toFixed(4));
console.log("Histogram:" +result.histogram.toFixed(4));
console.log("==============");
console.log("Signal:" +parseFloat(result.signal.valueOf()).toFixed(4));
console.log("Macd:" +parseFloat(result.macd.valueOf()).toFixed(4));
console.log("Histogram:" +parseFloat(result.histogram.valueOf()).toFixed(4));
```

But the output is the same:
```
Signal:0.2416
Macd:0.6178
Histogram:0.3762
==============
Signal:0.2416
Macd:0.6178
Histogram:0.3762
```

**Comment by bennycode** [comment]
Hi @Gonoshift, the MACD histogram is calculated as the MACD indicator minus the signal line. So if the signal line is different, then the histogram is also different. See: https://www.incrediblecharts.com/indicators/macd_histogram.php#macd_histogram-formula

The signal line can be configured by setting `indicator` and `signalInterval` in `MACDConfig`: https://github.com/bennycode/trading-signals/blob/v1.6.0/src/MACD/MACD.ts#L27-L31

My assumption is that the Python Tulip Indicators Library just uses a different signal line by default.

**Comment by Gonoshift** [comment]
@bennycode I'm not getting the "default signal line" point, the https://tulipindicators.org/macd example section, says:
`short period = 2, long period = 5, signal period = 9`

And that's why I've set it my parameters like this:
`const macd = new MACD({longInterval: 5, shortInterval: 2, signalInterval: 9, indicator: EMA});`

With the same explicit parameters it should output same result, doesn't it?

**Comment by bennycode** [comment]
I actually think it's a problem of the tulip indicators library.

The "macd" is short ema minus long ema.
The "macd signal" is signal ema minus macd.
The "macd histogram" is macd signal minus macd.

In their example calculation, the "macd" is "0.62" and the "macd signal" also is "0.62". That's why the histogram is "0.00". But I don't see how the "macd" and "macd signal" can be of the same values when using emas with short period = 2, long period = 5 and signal period = 9.

**Comment by Gonoshift** [comment]
@bennycode it could be... Or maybe the example is wrong and they used other parameters? 
Because if there was such an error in the Tupil library, I think that someone would have figured it out long before me...

**Comment by bennycode** [comment]
@Gonoshift, today I found out what's the difference...

In my library, the macd signal is updated directly from the start. In the "Tulip Indicators" example, the macd signal is only updated when there are at least 5 inputs (long period).

I adjusted my library accordingly and extended my MACD test to check histogram & signal line with the results from the Tulip Indicators example:
- https://github.com/bennycode/trading-signals/pull/219

You can get the update with **trading-signals v1.6.1**: 
- https://www.npmjs.com/package/trading-signals/v/1.6.1

If it fixes your issue, then I would be really happy about a star (⭐) from you:
- https://github.com/bennycode/trading-signals/stargazers

Best,
Benny

**Comment by VRciF** [comment]
Hi @bennycode,  
I was just experimenting with a few libraries and noticed, that macd differs greatly from the original `talib` c implementation.  
For comparision, a nodejs implementation, which uses talib and computes the same values as binance, can be found here: https://github.com/oransel/node-talib  

Another library producing the same values, as your lib does, is https://github.com/ixjb94/indicators.  
Thus, I'm not sure about the difference and if the values are wrongly calculated or maybe you follow another definition of macd.  
Maybe give node-talib a try and compare if the values actually differ  

As an example I used the values of ADAUSDT from 2024-10-02 00:00:00 to 2024-10-03 23:59:59 with 30m candles, see
https://api.binance.com/api/v3/klines?symbol=ADAUSDT&interval=30m&startTime=1727820000000&endTime=1727992799000

The close price has index 4 - see the docs for an example and details: https://binance-docs.github.io/apidocs/spot/en/#kline-candlestick-data

**Comment by VRciF** [comment]
Sorry, for highjacking this old thread.  
I'm not sure anymore if there is an issue.  

I mean, the values differ for sure for a given dataset, but regarding binance, I found this [thread](https://github.com/LupeCode/phpTraderNative/issues/11).  
The key parts are:
* Maybe, binance uses another EMA smoothing factor than the standard `2`, but instead uses `20`. Not sure if this is true
* And the values seem to better align if more historic candles are loaded beforehand. This is also what I have found myself. In the thread it is mentioned that with 1500 candles one should be on the safe side and be pretty accurate compared to binance

This kind of makes sense, as the EMA formula requires the previous EMA value. Thus if there are any differences, this might get worse. And I'm not sure about the convergence properties of EMA in general.
